models.py

```diff
@@ -280,19 +280,38 @@ def __get_distance_weights(d, w1, w0, sigma):
     return weights
 
 
-def __get_size_weights(sizes):
-    C = Variable(torch.sqrt(torch.Tensor([sizes.size()[1] * sizes.size()[2]])) / 2, requires_grad=False)
-    if torch.cuda.is_available():
-        C = C.cuda()
+def __get_size_weights(sizes, C):
     size_weights = C / sizes
     size_weights[sizes == 1] = 1
     return size_weights
 
 
-def get_loss_params(w0, sigma):
+def get_loss_params(w0, sigma, imsize):
     w0 = Variable(torch.Tensor([w0]), requires_grad=False)
     sigma = Variable(torch.Tensor([sigma]), requires_grad=False)
+    C = Variable(torch.sqrt(torch.Tensor([imsize[0] * imsize[1]])) / 2, requires_grad=False)
     if torch.cuda.is_available():
         w0 = w0.cuda()
         sigma = sigma.cuda()
-    return {'w0': w0, 'sigma': sigma}
+        C = C.cuda()
+    return {'w0': w0, 'sigma': sigma, 'C': C}
+
+
+def mixed_dice_cross_entropy_loss(output, target, dice_weight, ce_weight, ce_loss=None):
+    dice_target = target[:, 0, :, :].long()
+    ce_target = target
+    if ce_loss is None:
+        ce_loss = torch.nn.CrossEntropyLoss()
+        ce_target = dice_target
+    return dice_weight * dice_loss(output, dice_target) + ce_weight * ce_loss(output, ce_target)
+
+
+def dice_loss(output, target):
```

Comment:
@apyskir I think it is always a good idea to have some eps. But apparently having smooth could also help. As descibed herehttps://github.com/pytorch/pytorch/issues/1249 with this being a code snipped for smooth (no eps):

```
def dice_loss(input, target):
    smooth = 1.

    iflat = input.view(-1)
    tflat = target.view(-1)
    intersection = (iflat * tflat).sum()
    
    return 1 - ((2. * intersection + smooth) /
              (iflat.sum() + tflat.sum() + smooth))
```